pkg/controller/vitesscluster/reconcile_vtadmin.go

```diff
@@ -0,0 +1,373 @@
+	var vtctldServiceWebPort, vtctldServiceGrpcPort int32
+	for _, port := range vtctldService.Spec.Ports {
+		if port.Name == planetscalev2.DefaultWebPortName {
+			vtctldServiceWebPort = port.Port
+		}
+		if port.Name == planetscalev2.DefaultGrpcPortName {
+			vtctldServiceGrpcPort = port.Port
+		}
+	}
+
+	// Read the cell information
+	vtc := planetscalev2.VitessCell{}
+	err = r.client.Get(ctx, client.ObjectKey{
+		Namespace: vt.Namespace,
+		Name:      vitesscell.Name(vt.Name, cell.Name),
+	}, &vtc)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the vtgate service from the cell
+	vtgateService := corev1.Service{}
+	err = r.client.Get(ctx, client.ObjectKey{
+		Namespace: vtc.Namespace,
+		Name:      vtgate.ServiceName(vt.Name, cell.Name),
+	}, &vtgateService)
+	if err != nil {
+		return nil, err
+	}
+
+	// Find the IP address from the service. This is randomly assigned.
+	vtgateServiceIP := vtgateService.Spec.ClusterIP
+	// The grpc port should be set to the default value planetscalev2.DefaultGrpcPort,
+	// but since we have the service, we can just read it.
+	var vtgateServiceGrpcPort int32
+	for _, port := range vtgateService.Spec.Ports {
+		if port.Name == planetscalev2.DefaultGrpcPortName {
+			vtgateServiceGrpcPort = port.Port
+		}
+	}
+
+	// Variables to hold the key, value and secret name to use
+	discoveryKey := "discovery.json"
+	discoveryVal := fmt.Sprintf(`{
+    "vtctlds": [
+        {
+            "host": {
+                "fqdn": "%s:%d",
+                "hostname": "%s:%d"
+            }
+        }
+    ],
+    "vtgates": [
+        {
+            "host": {
+                "hostname": "%s:%d"
+            }
+        }
+    ]
+}`, vtctldServiceIP, vtctldServiceWebPort, vtctldServiceIP, vtctldServiceGrpcPort, vtgateServiceIP, vtgateServiceGrpcPort)
+	secretName := vtadmin.DiscoverySecretName(vt.Name, cell.Name)
+
+	// Create or update the secret
+	err = r.createOrUpdateSecret(ctx, vt, secretName, discoveryKey, discoveryVal)
+	if err != nil {
+		return nil, err
+	}
+
+	// return the secret source, which must align with the secret we created above
+	return &planetscalev2.SecretSource{
+		Name: secretName,
+		Key:  discoveryKey,
+	}, nil
+}
+
+func (r *ReconcileVitessCluster) createWebConfigSecret(ctx context.Context, vt *planetscalev2.VitessCluster, cell *planetscalev2.VitessCellTemplate, apiAddress string) (*planetscalev2.SecretSource, error) {
+	// Variables to hold the key, value and secret name to use
+	configKey := vtadmin.WebConfigFileName
+	configVal := fmt.Sprintf(`window.env = {
+     'REACT_APP_VTADMIN_API_ADDRESS': "%s",
+     'REACT_APP_FETCH_CREDENTIALS': "omit",
+     'REACT_APP_ENABLE_EXPERIMENTAL_TABLET_DEBUG_VARS': false,
+     'REACT_APP_BUGSNAG_API_KEY': "",
+     'REACT_APP_DOCUMENT_TITLE': "",
+     'REACT_APP_READONLY_MODE': %s,
+ };`, apiAddress, convertReadOnlyFieldToString(vt.Spec.VtAdmin.ReadOnly))
+	secretName := vtadmin.WebConfigSecretName(vt.Name, cell.Name)
```

Comment:
Some parameters of the config are not user controllable like `REACT_APP_ENABLE_EXPERIMENTAL_TABLET_DEBUG_VARS` is always set to false. Since we do not have a good way of discovering vttablets yet, this doesn't work. `http-tablet-url-tmpl` is not flexible to allow this right now. To enable this, we would have to setup services for all the vttablets and then encode them in the flag. Other complexities like new vttablets requiring change in config, etc will also have to be dealt with.
Similarly, `REACT_APP_FETCH_CREDENTIALS` is always set to `omit` because we aren't enabling CORS. If we did want to do that, then we would also need to ask the user to provide the hosts that are allowed to access the API server